### PR TITLE
Upgrade To THREE 0.166.0, upgrade to Expo SDK 51

### LINKED ISSUE #164: [Expo Three] Snacks / builds are broken?
Hi everyone,

Currently trying to get Expo and Expo Three working.

When I open: https://snack.expo.io/@bacon/three

I see this: 
image

Console:
image

Looks like a Cloudfront error?

iOS / Android also does not work, any idea? I have no Chrome extensions enabled and also tried in Edge and Chrome incognito mode.

### PR DESCRIPTION
**Pull request by markrickert**
This PR is the culmination of a few weeks of work and experimentation around the `expo-three` and `three.js` ecosystems. @Jpoliachik and I were able to get many more examples running in the example app and lots of upgrades to bring this library back into line with the latest expo standards and SDK.

Here are the updates in no particular order:

1. The example app is set up to use the source code of `expo-three` instead of the build directory, making developing on the example app much easier without the build service running.
2. Updated example app to Expo SDK 51 and added section in readme for how to run the example app.
3. Updated `expo-three` to use the latest `expo-module-scripts`
4. Updated to the latest version of THREE (0.166.0) and fixed incompatibilities and deprecations in the examples. There was a [global lighting system update](https://discourse.threejs.org/t/updates-to-lighting-in-three-js-r155/53733) in r155 so some work had to be done to get the lighting system to get the examples working again. 
5. Added a rudimentary stats panel to the example
6. Fixed the rudimentary file extension matching (it was just looking at the end of the url) with a more robust url matcher that takes parameters in the url into account. With tests.
7. Added orbit controls demo with `expo-three-orbit-controls` (had to patch because it still relies on an old version of THREE).
8. Added lots more demos

The example app works on android and web, but I haven't tested on iOS because the simulator doesn't support web-gl.

| iOS | Android | Web |
| --- | --- | --- |
| 🤷‍♂️ | ✅ | ✅ |

All tests pass (including the new ones!)
Linter passes

Please let me know what else needs to be done to get this merged!

This PR will Close: #164, #199, #212, #238, #270, #287, #297, #298, & #299

Thanks to @Jpoliachik for his help in getting this branch working and for porting over some THREE web examples over from the web examples.

### CODE REVIEW

**Comment by andrija29** [comment]
Great work guys! Let's get this merged please @keith-kurak

**Comment by joekendal** [comment]
thanks

**Comment by adavidson21** [comment]
This is great stuff! Would love to see it merged.

**Comment by sebastian-nowak** [comment]
Any updates on this? This really needs to be merged, `expo-three` is too far behind the current `three` version

**Comment by Benmidi** [comment]
Ooof, I really need this too.

**Comment by markrickert** [comment]
I've resolved all the merge conflicts and upgraded ThreeJS to the latest `0.165.0`.

I also added an example of using the `GCodeLoader` 🚢 

Screenshot 2024-06-12 at 5 50 56 PM

**Comment by davidlumley** [comment]
Any updates on what is required for this to be merged?

**Comment by joekendal** [comment]
Lgtm

**Comment by joekendal** [comment]
@EvanBacon @tsapeta @lukmccall

**Comment by joekendal** [comment]
@markrickert I have tested your PR to address #297 

**package.json**
```json
"expo-three": "github:markrickert/expo-three.git#dev",
```

It does not seem to resolve the error: *Failed to getSize of file:///var/mobile/Containers/Data/Application/...*

**Comment by Josema** [comment]
I am having some problems instaling the branch repository.

`Unable to resolve "expo-three" from "src/components/App.js"`

```
"expo": "~51.0.17",
"expo-three": "markrickert/expo-three#dev",
"three": "^0.166.0"
```

I also tried installing it this way:
`"expo-three": "git+https://github.com/markrickert/expo-three.git#dev",` 

But it says:
git@github.com: Permission denied (publickey).

File: example/metro.config.js [hunk 1]
```diff
@@ -1,47 +1,40 @@
...
 
-    extraNodeModules: modules.reduce((acc, name) => {
-      acc[name] = path.join(__dirname, 'node_modules', name);
-      return acc;
-    }, {}),
-  },
+config.resolver.assetExts.push('png');
+config.resolver.sourceExts = config.resolver.sourceExts.filter(
+  (ext) => ext !== 'png'
+);
```
Comment (tsapeta) [inline example/metro.config.js:29]: Isn't png already an asset extension?

File: .eslintrc.js [hunk 2]
```diff
@@ -1,2 +1,5 @@
-// @generated by expo-module-scripts
-module.exports = require('expo-module-scripts/eslintrc.base.js');
+module.exports = {
+  root: true,
+  extends: ['universe/native', 'universe/web'],
+  ignorePatterns: ['build'],
+};
```
Comment (tsapeta) [inline .eslintrc.js:5]: Why do we need a custom config?

File: example/plugins/withFlipperDisabled.ts
Comment (tsapeta) [inline example/plugins/withFlipperDisabled.ts:1]: Expo doesn't support Flipper and it's also officially deprecated. As far as I remember the app template doesn't include it at all, so why do we need a plugin that disables it?

**Review by tsapeta — COMMENTED** [review]
Hey there!
First of all, I'm sorry that this PR has been left alone 😅 I think this happened mostly due to a large scope of changes, including ones that are probably unnecessary and/or could be split into smaller pieces so it's easier for us to review it.

I did the first pass of review, mostly just questions so I can better understand the motivation behind those changes.

I really appreciate the time and effort you guys took to make this!

**Comment by markrickert** [comment]
@tsapeta Thanks for the review. I'm new to working on expo modules and i'm just trying to push this library forward. I've addressed your comments in a new commit and i'll comment or resolve them here on github too.

File: package.json [hunk 3]
```diff
@@ -17,10 +17,14 @@
     "build": "expo-module build",
     "clean": "expo-module clean",
     "test": "expo-module test",
-    "lint": "expo-module lint",
+    "lint": "eslint src --ext .js,.ts,.tsx",
```
[the remark this answers is not in the document]
Comment (markrickert) [reply package.json]: You're right. I switched it back. I did this while trying to figure out the exact versions of eslint and other dependencies we needed to get the linter passing.

File: example/webpack.config.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+const createConfigAsync = require('@expo/webpack-config');
```
[the remark this answers is not in the document]
Comment (markrickert) [reply example/webpack.config.js]: You're right. I deleted this file.

File: package.json [hunk 5]
```diff
@@ -46,24 +50,41 @@
...
+    "eslint-plugin-promise": "6.0.0",
+    "eslint-plugin-react": "7.30.0",
+    "eslint-plugin-react-native": "4.0.0",
+    "eslint-plugin-reactotron": "^0.1.2",
+    "expo": "^51.0.17",
+    "expo-asset": "~9.0.2",
+    "expo-file-system": "~16.0.9",
+    "expo-gl": "~13.6.0",
+    "expo-module-scripts": "3.5.1",
+    "expo-modules-core": "1.12.11",
```
[the remark this answers is not in the document]
Comment (markrickert) [reply package.json]: Understood. I removed it without consequence. Thanks for the tip!

File: example/plugins/withBuildProperties.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
+const NEW_ARCH_ENABLED = false;
+
+/**
+ *
+ */
```
[the remark this answers is not in the document]
Comment (markrickert) [reply example/plugins/withBuildProperties.ts]: Deleted this file and moved the build properties to the main app config file.

File: package.json [hunk 7]
```diff
@@ -46,24 +50,41 @@
...
+    "eslint-plugin-n": "^15.0.0",
+    "eslint-plugin-prettier": "^4.0.0",
+    "eslint-plugin-promise": "6.0.0",
+    "eslint-plugin-react": "7.30.0",
+    "eslint-plugin-react-native": "4.0.0",
+    "eslint-plugin-reactotron": "^0.1.2",
+    "expo": "^51.0.17",
+    "expo-asset": "~9.0.2",
+    "expo-file-system": "~16.0.9",
+    "expo-gl": "~13.6.0",
```
[the remark this answers is not in the document]
Comment (markrickert) [reply package.json]: I was able to get rid of `expo-gl` but all the others need to b there for tests to pass :D

File: example/metro.config.js (hunk 1 above)
Comment (markrickert) [reply example/metro.config.js]: Yup, this was left in accidentally. Resolved.

File: .eslintrc.js (hunk 2 above)
Comment (markrickert) [reply .eslintrc.js]: Turns out we didnt! I deleted the file and the next time i ran `yarn` it was recreated from the template. linting still works 🥳

File: example/plugins/withFlipperDisabled.ts
Comment (markrickert) [reply example/plugins/withFlipperDisabled.ts:1]: Deleted!

**Comment by whygolfalan** [comment]
How is this looking? Could really use it!

**Review by tsapeta — APPROVED** [review]
Looking good now, thank you! ❤️

**Comment by tsapeta** [comment]
It's been published in version `8.0.0` 🎉 Could someone confirm if everything is working fine at first glance? I'll then promote this version to `latest` on npm.

**Comment by Josema** [comment]
> It's been published in version `8.0.0` 🎉 Could someone confirm if everything is working fine at first glance? I'll then promote this version to `latest` on npm.

It is working for me. 👍🎉

```
"expo": "~51.0.18",
"expo-gl": "~14.0.2",
"expo-three": "^8.0.0",
"three": "^0.166.0"
 ```

**Comment by markrickert** [comment]
@tsapeta I just checked in a new app, adding `expo-gl`, `expo-three`, and `three` at the versions specified in the comment above and it loaded in the Gcode example perfectly!

Thanks for getting this over the finish line! 🥳

**Pull request merged**